src/System.Private.CoreLib/shared/System/Resources/ResourceReader.cs

```diff
@@ -743,7 +758,81 @@ private object _LoadObjectV2(int pos, out ResourceTypeCode typeCode)
 
             // Normal serialized objects
-            throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
+            int typeIndex = typeCode - ResourceTypeCode.StartOfUserTypes;
+            return DeserializeObject(typeIndex);
+        }
+
+        // Helper method to deserialize a type
+        private Object DeserializeObject(int typeIndex)
+        {
+            if (!_permitDeserialization)
+            {
+                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
+            }
+
+            if (_binaryFormatter == null)
+            {
+                InitializeBinaryFormatter();
+            }
+
+            Type type = FindType(typeIndex);
+ 
+            // Ensure that the object we deserialized is exactly the same
+            // type of object we thought we should be deserializing. 
+            Object graph;
+            graph = s_deserializeMethod(_binaryFormatter, _store.BaseStream);
+            
+            // This check is about correctness, not security at this point.
+            if (graph.GetType() != type)
+                throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResType_SerBlobMismatch, type.FullName, graph.GetType().FullName));
+ 
+            return graph;
+        }
+
+        private void InitializeBinaryFormatter()
+        {
+            if (s_binaryFormatterType == null)
+            {
+                Type binaryFormatterType = Type.GetType(
+                    "System.Runtime.Serialization.Formatters.Binary.BinaryFormatter, System.Runtime.Serialization.Formatters, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
+                    throwOnError: false);
+                if (binaryFormatterType != null)
+                {
+                    ConstructorInfo binaryFormatterCtor = binaryFormatterType.GetConstructor(new Type[0]);
+                    MethodInfo binaryFormatterDeserialize = binaryFormatterType.GetMethod("Deserialize", new Type[] { typeof(Stream) });
+
+                    s_binaryFormatterType = binaryFormatterType;
+                    // create an unbound delegate that can accept a BinaryFormatter instance as object
+                    s_deserializeMethod = (Func<object, Stream, object>)typeof(ResourceReader)
+                                           .GetMethod(nameof(CreateUntypedDelegate), BindingFlags.NonPublic | BindingFlags.Static)
+                                           .MakeGenericMethod(binaryFormatterType)
+                                           .Invoke(null, new object[] { binaryFormatterDeserialize });
+                }
+                else
+                {
+                    // set a sentinel to indicate that BinaryFormatter wasn't found
+                    s_binaryFormatterType = typeof(object);
+                    s_deserializeMethod = null;
+                }
+            }
+
+            if (s_binaryFormatterType != null && s_deserializeMethod == null)
```

Comment:
This has race condition. We can get here while a second thread could have successfully initialized `s_binaryFormatterType` and not initialized `s_deserializeMethod` yet.

I would just check `s_deserializeMethod  != null` here and above since `s_deserializeMethod` is set last. I would not worry about limiting amount of work done on the exception throwing path by using the sentinel.

Reply:
Alternatively, this could just use something like:
```C#
private static readonly Lazy<(Type, Func<object, Stream, object>)> s_binaryFormatter = new Lazy<(Type, Func<object, Stream, object>)>(() =>
{
    ... // initialization logic
    return (binaryFormatterType, deserializeMethod);
});
```
and let it handle the safety.